File: crossdata-driver/src/main/scala/com/stratio/crossdata/driver/querybuilder/Select.scala
/*
 * Licensed to STRATIO (C) under one or more contributor license agreements.
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership.  The STRATIO (C) licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.stratio.crossdata.driver.querybuilder

/**
 * Select query builder.
 */
class Select extends Query {

  /**
   * The selected columns.
   */
  var selection : Option[Selection] = None

  /**
   * The target table in the from clause.
   */
  var from : Option[String] = None

  /**
   * The join clause.
   */
  var join : Option[Join] = None

  /**
   * The where clause.
   */
  var where : Option[Where] = None

  /**
   * The window clause.
   */
  var window : Option[String] = None

  /**
   * The group by clause.
   */
  var groupBy : Option[String] = None

  /**
   * The order by clause.
   */
  var orderBy : Option[String] = None

  /**
   * The limit clause.
   */
  var limit : Option[Long] = None

  /**
   * Class constructor.
   * @param columnName The name of the column to be selected.
   */
  def this(columnName : String){
    this()
    selection = Option(new Selection(List(columnName)))
  }

  /**
   * Class constructor.
   * @param columnNames A sequence of string column names.
   */
  def this(columnNames : Seq[String]){
    this()
    selection = Option(new Selection(columnNames.toList))
  }

  /**
   * Class constructor.
   * @param selectedColumns The Selection with the required columns.
   */
  def this(selectedColumns : Selection){
    this()
    selection = Option(selectedColumns)
  }

  /**
   * Add a from clause to the current query.
   * @param tableName The target table name.
   * @return The current Select object.
   */
  def from(tableName : String) : Select = {
    from = Option(tableName)
    this
  }

  /**
   * Add a join clause to the current query.
   * @param tableName The target table to perform the join.
   * @return A default Join.
   */
  def join(tableName : String) : Join = {
    innerJoin(tableName)
  }

  /**
   * Add an inner join clause to the current query.
   * @param tableName The target table to perform the join.
   * @return An inner Join
   */
  def innerJoin(tableName: String) : Join = {
    join = Option(new Join(tableName, "INNER", this))
    join.get
  }

  /**
   * Add a window clause to the current query.
   * @param window The window.
   * @return The current Select object.
   */
  def withWindow(window: String) : Select = {
    this.window = Option(window)
    this
  }

  /**
   * Add a where clause to the current query.
   * @param clause The where clause.
   * @return A Where object.
   */
  def where(clause : String) : Where = {
    where = Option(new Where(clause, this))
    where.get
  }

  /**
   * Add a group by clause to the current query.
   * @param clause The group by clause.
   * @return The current Select object.
   */
  def groupBy(clause : String) : Select = {
    groupBy = Option(clause)
    this
  }

  /**
   * Add an order by clause to the current query.
   * @param clause The order by clause.
   * @return The current Select object.
   */
  def orderBy(clause : String) : Select = {
    orderBy = Option(clause)
    this
  }

  /**
   * Add a limit clause to the current query.
   * @param queryLimit The limit.
   * @return The current Select object.
   */
  def limit(queryLimit : Long) : Select = {
    limit = Option(queryLimit)
    this
  }

  override def toString() : String = {
    val sb : StringBuilder = new StringBuilder("SELECT ");
    if(selection.nonEmpty){
      sb.append(selection.get.asString)
    }

    if(from.nonEmpty){
      sb.append(" FROM ").append(from.get)
    }

    if(window.nonEmpty){
      sb.append(" WITH WINDOW ").append(window.get)
    }

    if(join.nonEmpty){
      sb.append(join.get.asString)
    }

    if(where.nonEmpty){
      sb.append(" WHERE ").append(where.get.asString)
    }

    if(orderBy.nonEmpty){
      sb.append(" ORDER BY ").append(orderBy.get)
    }

    if(groupBy.nonEmpty){
      sb.append(" GROUP BY ").append(groupBy.get)
    }

    if(limit.nonEmpty){
      sb.append(" LIMIT ").append(limit.get)
    }

    sb append ";"

    sb mkString
  }
}

File: crossdata-common/src/main/java/com/stratio/crossdata/common/logicalplan/Select.java
/*
 * Licensed to STRATIO (C) under one or more contributor license agreements.
 * See the NOTICE file distributed with this work for additional information
 * regarding copyright ownership.  The STRATIO (C) licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.stratio.crossdata.common.logicalplan;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import com.stratio.crossdata.common.metadata.Operations;
import com.stratio.crossdata.common.data.ColumnName;
import com.stratio.crossdata.common.metadata.ColumnType;
import com.stratio.crossdata.common.statements.structures.Selector;

/**
 * Select operator that specifies the list of columns that should be returned with their expected name. Notice that this
 * operator is applied to limit the number of columns returned and to provide alias support.
 */
public class Select extends TransformationStep {

    /**
     * Map of columns associating the name given in the Project logical steps with the name expected in the results.
     */
    private final Map<Selector, String> columnMap;

    /**
     * Map of {@link com.stratio.crossdata.common.metadata.ColumnType} associated with each column.
     */
    private Map<String, ColumnType> typeMap;

    /**
     * Map with the types of the selectors.
     */
    private Map<Selector, ColumnType> typeMapFromColumnName;

    /**
     * Class constructor.
     * 
     * @param operation
     *            The operation to be applied.
     * @param columnMap
     *            Map of columns associating the name given in the Project logical steps with the name expected in the
     *            result.
     * @param typeMap
     *            The mapping of column types.
     * @param typeMapFromColumnName  The types of selectors.
     */
    public Select(Operations operation, Map<Selector, String> columnMap, Map<String, ColumnType> typeMap,
            Map<Selector, ColumnType> typeMapFromColumnName) {
        super(operation);
        this.columnMap = columnMap;
        this.typeMap = typeMap;
        this.typeMapFromColumnName = typeMapFromColumnName;
    }

    public Map<Selector, String> getColumnMap() {
        return columnMap;
    }

    public Map<String, ColumnType> getTypeMap() {
        return typeMap;
    }

    public Map<Selector, ColumnType> getTypeMapFromColumnName() {
        return typeMapFromColumnName;
    }

    public void setTypeMap(Map<String, ColumnType> typeMap) {
        this.typeMap = typeMap;
    }

    public void setTypeMapFromColumnName(Map<Selector, ColumnType> typeMapFromColumnName) {
        this.typeMapFromColumnName = typeMapFromColumnName;
    }

    /**
     * Get the selectors that must be returned as a result of executing the query
     * in the expected order.
     * @return A list of {@link com.stratio.crossdata.common.statements.structures.Selector}.
     */
    public List<Selector> getOutputSelectorOrder(){
        List<Selector> result = new ArrayList<>();
        for(Selector selector : columnMap.keySet()){
            result.add(selector);
        }
        return result;
    }

    /**
     * @deprecated
     * Get the columns in the expected order. With the introduction of function support
     * this method is marked as Deprecated in favour of #getOutputSelectorOrder().
     * @return A list of {@link com.stratio.crossdata.common.data.ColumnName}.
     */
    @Deprecated
    public List<ColumnName> getColumnOrder(){
        List<ColumnName> results = new ArrayList<>();
        for(Selector selector: columnMap.keySet()){
            results.add(selector.getColumnName());
        }
        return results;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder("SELECT (");
        Iterator<Map.Entry<Selector, String>> it = columnMap.entrySet().iterator();
        Map.Entry<Selector, String> entry;
        while (it.hasNext()) {
            entry = it.next();
            sb.append(entry.getKey().getColumnName().getQualifiedName()).append(" AS ").append(entry.getValue());
            if (it.hasNext()) {
                sb.append(", ");
            }
        }
        sb.append(")");
        return sb.toString();
    }
}
